File: framework/languages/node/e2e/RegistrationCodec/Server/Main/Handlers/registration-handlers.ts
import { Injectable } from '@nestjs/common';
import {
  zlinkRequestHandler,
  zlinkSendHandler
} from '@zlink-systems/nestjs';
import type {
  ZLinkMessageContext,
  ZLinkRequestHandler,
  ZLinkSendHandler
} from '@zlink-systems/framework';
import { PacketNames, type EchoMsg, type EchoRes, type EchoReq } from '../../../Shared/messages';
import { EvidenceStore } from '../Infrastructure/evidence-store';

@zlinkRequestHandler('auto', PacketNames.echoAutoReq)
@Injectable()
export class EchoAutoRequestHandler implements ZLinkRequestHandler<EchoReq, EchoRes> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: EchoReq, context: ZLinkMessageContext): Promise<EchoRes> {
    this.evidence.add(`echo-request|variant=auto|value=${request.value}|content=${context.contentType}`);
    return { value: `echo:${request.value}`, contentType: context.contentType ?? '<null>' };
  }
}

@zlinkSendHandler('auto', PacketNames.echoAutoMsg)
@Injectable()
export class EchoAutoCommandHandler implements ZLinkSendHandler<EchoMsg> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(message: EchoMsg, context: ZLinkMessageContext): Promise<void> {
    this.evidence.add(`echo-command|variant=auto|id=${message.commandId}|value=${message.value}|content=${context.contentType}`);
  }
}

@zlinkRequestHandler('attr', PacketNames.echoAttrReq)
@Injectable()
export class EchoAttrRequestHandler implements ZLinkRequestHandler<EchoReq, EchoRes> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: EchoReq, context: ZLinkMessageContext): Promise<EchoRes> {
    this.evidence.add(`echo-request|variant=attr|value=${request.value}|content=${context.contentType}`);
    return { value: `echo:${request.value}`, contentType: context.contentType ?? '<null>' };
  }
}

@zlinkSendHandler('attr', PacketNames.echoAttrMsg)
@Injectable()
export class EchoAttrCommandHandler implements ZLinkSendHandler<EchoMsg> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(message: EchoMsg, context: ZLinkMessageContext): Promise<void> {
    this.evidence.add(`echo-command|variant=attr|id=${message.commandId}|value=${message.value}|content=${context.contentType}`);
  }
}

@Injectable()
export class EchoManualRequestHandler implements ZLinkRequestHandler<EchoReq, EchoRes> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: EchoReq, context: ZLinkMessageContext): Promise<EchoRes> {
    this.evidence.add(`echo-request|variant=manual|value=${request.value}|content=${context.contentType}`);
    return { value: `echo:${request.value}`, contentType: context.contentType ?? '<null>' };
  }
}

@Injectable()
export class EchoManualCommandHandler implements ZLinkSendHandler<EchoMsg> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(message: EchoMsg, context: ZLinkMessageContext): Promise<void> {
    this.evidence.add(`echo-command|variant=manual|id=${message.commandId}|value=${message.value}|content=${context.contentType}`);
  }
}

@Injectable()
export class DuplicateEchoRequestHandler implements ZLinkRequestHandler<EchoReq, EchoRes> {
  async handle(request: EchoReq): Promise<EchoRes> {
    return { value: request.value, contentType: 'duplicate' };
  }
}

File: framework/languages/dotnet/e2e/RegistrationCodec/Server/Main/Handlers/RegistrationHandlers.cs
using RegistrationCodec.Server.Main.Infrastructure;
using RegistrationCodec.Shared;
using Zlink.Framework.Contracts.Handlers;

namespace RegistrationCodec.Server.Main.Handlers;

[ZLinkHandlerGroup("auto")]
internal sealed class EchoAutoRequestHandler(EvidenceStore evidence)
    : IZLinkRequestHandler<EchoAutoReq, EchoRes>
{
    public ValueTask<EchoRes> HandleAsync(EchoAutoReq request, IZLinkMessageContext context,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"echo-request|variant=auto|value={request.Value}|content={context.ContentType}");
        return ValueTask.FromResult(new EchoRes($"echo:{request.Value}", context.ContentType ?? "<null>"));
    }
}

[ZLinkHandlerGroup("auto")]
internal sealed class EchoAutoCommandHandler(EvidenceStore evidence)
    : IZLinkSendHandler<EchoAutoMsg>
{
    public ValueTask HandleAsync(EchoAutoMsg message, IZLinkMessageContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add(
            $"echo-command|variant=auto|id={message.CommandId}|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }
}

[ZLinkHandlerGroup("attr")]
internal sealed class AttributeHandlers(EvidenceStore evidence)
    : IZLinkRequestHandler<EchoAttrReq, EchoRes>,
      IZLinkSendHandler<EchoAttrMsg>
{
    [ZLinkRequest(PacketName = "EchoAttr")]
    public EchoRes Request(EchoAttrReq request, IZLinkMessageContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"echo-request|variant=attr|value={request.Value}|content={context.ContentType}");
        return new EchoRes($"echo:{request.Value}", context.ContentType ?? "<null>");
    }

    [ZLinkSend(PacketName = "EchoAttrMsg")]
    public ValueTask Send(EchoAttrMsg message, IZLinkMessageContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add(
            $"echo-command|variant=attr|id={message.CommandId}|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }

    ValueTask<EchoRes> IZLinkRequestHandler<EchoAttrReq, EchoRes>.HandleAsync(
        EchoAttrReq request,
        IZLinkMessageContext context,
        CancellationToken cancellationToken) =>
        ValueTask.FromResult(Request(request, context, cancellationToken));

    ValueTask IZLinkSendHandler<EchoAttrMsg>.HandleAsync(
        EchoAttrMsg message,
        IZLinkMessageContext context,
        CancellationToken cancellationToken) =>
        Send(message, context, cancellationToken);
}

internal sealed class EchoManualRequestHandler(EvidenceStore evidence)
    : IZLinkRequestHandler<EchoManualReq, EchoRes>
{
    public ValueTask<EchoRes> HandleAsync(EchoManualReq request, IZLinkMessageContext context,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"echo-request|variant=manual|value={request.Value}|content={context.ContentType}");
        return ValueTask.FromResult(new EchoRes($"echo:{request.Value}", context.ContentType ?? "<null>"));
    }
}

internal sealed class EchoManualCommandHandler(EvidenceStore evidence)
    : IZLinkSendHandler<EchoManualMsg>
{
    public ValueTask HandleAsync(EchoManualMsg message, IZLinkMessageContext context,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add(
            $"echo-command|variant=manual|id={message.CommandId}|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }
}

internal sealed class DuplicateEchoRequestHandler
    : IZLinkRequestHandler<EchoManualReq, EchoRes>
{
    public ValueTask<EchoRes> HandleAsync(EchoManualReq request, IZLinkMessageContext context,
        CancellationToken cancellationToken)
    {
        _ = context;
        cancellationToken.ThrowIfCancellationRequested();
        return ValueTask.FromResult(new EchoRes(request.Value, "duplicate"));
    }
}
